File: backend/gemini-proxy/src/main/java/com/tranphuc8a/gemini_proxy/application/usecases/ConversationUseCase.java
package com.tranphuc8a.gemini_proxy.application.usecases;

import com.tranphuc8a.gemini_proxy.application.ports.input.ConversationInputPort;
import com.tranphuc8a.gemini_proxy.application.ports.output.ConversationOutputPort;
import com.tranphuc8a.gemini_proxy.application.ports.output.MessageOutputPort;
import com.tranphuc8a.gemini_proxy.domain.exceptions.user.ResourceNotFoundException;
import com.tranphuc8a.gemini_proxy.domain.models.Conversation;
import com.tranphuc8a.gemini_proxy.domain.models.Message;
import com.tranphuc8a.gemini_proxy.domain.vo.request.ConversationUpdatingRequest;
import com.tranphuc8a.gemini_proxy.domain.vo.response.ResponseResult;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ConversationUseCase implements ConversationInputPort {

    public static final int LATEST_MESSAGE_COUNT = 10;
    public static final String CONVERSATION_NEW_NAME = "New Conversation";

    private final ConversationOutputPort conversationOutputPort;

    private final MessageOutputPort messageOutputPort;

    @Override
    public ResponseEntity<Object> getConversationDetail(String conversationId) {
        Conversation conversation = conversationOutputPort.getById(conversationId);
        if (conversation == null) {
            throw new ResourceNotFoundException("Conversation not found");
        }
        conversation.setMessages(
                conversation.getMessages().stream().limit(LATEST_MESSAGE_COUNT).toList()
        );
        return ResponseEntity.ok(ResponseResult.from(
                conversation, "Get conversation detail successfully"));
    }

    @Override
    public ResponseEntity<Object> getConversationList(Pageable pageable) {
        Page<Conversation> conversationPage = conversationOutputPort.getAll(pageable);
        conversationPage.forEach(conversation -> conversation.setMessages(List.of()));
        return ResponseEntity.ok(ResponseResult.from(
                conversationPage, "Get conversation list successfully"));
    }

    @Override
    public ResponseEntity<Object> getConversationMessages(String conversationId, Pageable pageable) {
        Conversation conversation = conversationOutputPort.getById(conversationId);
        if (conversation == null) {
            throw new ResourceNotFoundException("Conversation not found");
        }
        Page<Message> messagePage = messageOutputPort.getListByConversation(conversationId, pageable);
        messagePage.forEach(message -> message.setConversation(null));
        return ResponseEntity.ok(ResponseResult.from(
                messagePage, "Get conversation messages successfully"));
    }

    @Override
    public ResponseEntity<Object> createConversation() {
        Conversation conversation = Conversation.builder()
                .id(null)
                .name(CONVERSATION_NEW_NAME)
                .build();
        conversation = conversationOutputPort.save(conversation);
        return ResponseEntity.ok(ResponseResult.from(
                conversation, "Create conversation successfully"));
    }

    @Override
    public ResponseEntity<Object> updateConversation(ConversationUpdatingRequest conversationUpdatingRequest) {
        Conversation conversation = conversationOutputPort.getById(conversationUpdatingRequest.getConversationId());
        if (conversation == null) {
            throw new ResourceNotFoundException("Conversation not found");
        }
        conversation.setName(conversationUpdatingRequest.getName());
        conversation = conversationOutputPort.save(conversation);
        return ResponseEntity.ok(ResponseResult.from(
                conversation, "Update conversation successfully"));
    }

    @Override
    public ResponseEntity<Object> deleteConversation(String conversationId) {
        Conversation conversation = conversationOutputPort.getById(conversationId);
        if (conversation == null) {
            throw new ResourceNotFoundException("Conversation not found");
        }
        conversationOutputPort.delete(conversation);
        return ResponseEntity.ok(ResponseResult.from(
                conversation, "Delete conversation successfully"));
    }
}

File: backend/fastapi/src/application/usecases/conversation_usecase.py
from typing import Optional
from src.domain.enums.enums import ERole
from src.domain.models.message_domain import MessageDomain
from src.application.ports.input.conversation_input_port import ConversationInputPort
from src.application.ports.output.conversation_output_port import ConversationOutputPort
from src.application.ports.output.message_output_port import MessageOutputPort
from src.domain.utils.utils import generate_unique_id, get_current_timestamp
from src.domain.vo.message_response import MessageResponse
from src.domain.vo.conversation_update_request import ConversationUpdateRequest
from src.domain.vo.conversation_response import ConversationResponse
from src.domain.vo.list_response import ListResponse
from src.domain.models.conversation_domain import ConversationDomain
from fastapi import HTTPException


class ConversationUseCase(ConversationInputPort):
    LATEST_MESSAGE_COUNT = 10
    CONVERSATION_NEW_NAME = "New Conversation"

    def __init__(self, conversation_repo: ConversationOutputPort, message_repo: MessageOutputPort):
        self.conversation_repo = conversation_repo
        self.message_repo = message_repo

    async def get_conversation_detail(self, conversation_id: str) -> ConversationResponse:
        conv = await self.conversation_repo.get_by_id(conversation_id)
        if conv is None:
            raise HTTPException(status_code=404, detail="Conversation not found")
        
        # fetch latest messages explicitly to avoid async lazy-loading issues
        messages, _ = await self.message_repo.get_list_by_conversation(conversation_id, limit=self.LATEST_MESSAGE_COUNT, after=None, order="desc")
        conv.messages = messages
        return ConversationResponse.from_domain(conv)

    async def get_conversation_list(self, 
                                    after: Optional[str] = None, 
                                    limit: int = 10, 
                                    order: Optional[str] = "desc") -> ListResponse[ConversationResponse]:
        """Return a paginated ListResponse of conversations.

        - order: 'asc' or 'desc' (by created_at)
        - after: id cursor (optional)
        - offset: number of items to skip after the cursor
        - limit: max items to return
        """
        items, has_more = await self.conversation_repo.get_all(limit=limit, after=after, order=order if order else "desc")
        # convert to response VO list
        data = [ConversationResponse.from_domain(c) for c in items]
        first_id = data[0].id if data else None
        last_id = data[-1].id if data else None
        return ListResponse[ConversationResponse](data=data, first_id=first_id, last_id=last_id, has_more=has_more)

    async def get_conversation_messages(self, 
                                        conversation_id: str, 
                                        after: Optional[str] = None, 
                                        limit: int = 10, 
                                        order: Optional[str] = "desc") -> ListResponse[MessageResponse]:
        conv = await self.conversation_repo.get_by_id(conversation_id)
        if conv is None:
            raise HTTPException(status_code=404, detail="Conversation not found")
        messageList, has_more = await self.message_repo\
            .get_list_by_conversation(conversation_id, after=after, limit=limit, order=order if order else "desc")
        data = [MessageResponse.from_domain(m) for m in messageList]
        first_id = data[0].id if data else None
        last_id = data[-1].id if data else None
        return ListResponse[MessageResponse](
            data=data, 
            first_id=first_id, 
            last_id=last_id, 
            has_more=has_more
        )

    async def create_conversation(self) -> ConversationResponse:
        conv = ConversationDomain(
            id=generate_unique_id("conv"), 
            name=self.CONVERSATION_NEW_NAME,
            created_at=get_current_timestamp(),
            updated_at=None
        )
        saved = await self.conversation_repo.save(conv)
        return ConversationResponse.from_domain(saved)

    async def update_conversation(self, request: ConversationUpdateRequest) -> ConversationResponse:
        conv = await self.conversation_repo.get_by_id(request.id)
        if conv is None:
            raise HTTPException(status_code=404, detail="Conversation not found")
        conv.name = request.name
        conv.updated_at = get_current_timestamp()
        updated = await self.conversation_repo.save(conv)
        return ConversationResponse.from_domain(updated)

    async def delete_conversation(self, conversation_id: str) -> bool:
        conv = await self.conversation_repo.get_by_id(conversation_id)
        if conv is None:
            raise HTTPException(status_code=404, detail="Conversation not found")
        await self.conversation_repo.delete(conv)
        return True

    async def post_message(self, conversation_id: str, content: str) -> MessageResponse:
        conv = await self.conversation_repo.get_by_id(conversation_id)
        if conv is None:
            raise HTTPException(status_code=404, detail="Conversation not found")
        
        message = MessageDomain(
            id=generate_unique_id("msg"),
            conversation_id=conversation_id,
            role=ERole.USER,
            content=content,
            created_at=get_current_timestamp(),
        )
        saved_message = await self.message_repo.save(message)
        return MessageResponse.from_domain(saved_message)
    